packages/go_router/lib/src/route.dart

```diff
@@ -432,8 +432,13 @@ class GoRoute extends RouteBase {
 
   // TODO(chunhtai): move all regex related help methods to path_utils.dart.
   /// Match this route against a location.
-  RegExpMatch? matchPatternAsPrefix(String loc) =>
-      _pathRE.matchAsPrefix(loc) as RegExpMatch?;
+  RegExpMatch? matchPatternAsPrefix(String loc) {
+    if (!loc.startsWith('/') && _pathRE.hasMatch('/$loc')) {
+      loc = '/$loc';
+    }
+    return _pathRE.matchAsPrefix('/$loc') as RegExpMatch? ??
+        _pathRE.matchAsPrefix(loc) as RegExpMatch?;
```

Comment:
I believe it would ultimately be better to require all routes to start with `/` so there is not this type of weird stuff going on. Here we can't predict whether or not `loc` starts with `/` or not.